pylana/utils.py

```diff
@@ -62,40 +62,36 @@ def create_event_semantics_from_df(df: pd.DataFrame, time_format: str = "yyyy-MM
     time stamps need to have the same time format. For an overview over time stamp formats
     see https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatter.html#patterns
     """
-
-    dct_bare = [
-        {'name': 'Case_ID', 'semantic': 'Case ID', 'format': None},
-        {'name': 'Action', 'semantic': 'Action', 'format': None}
-    ]
-
-    bares = ['Case_ID', 'Action']
-
-    dct_bare += [{
-        'name': col,
-        'semantic': col,
-        'format': time_format}
-        for col in df.columns.intersection(['Start', 'Complete'])
-        if col not in bares
-    ]
-
-    dct_bare += [{
-        'name': col,
-        'semantic': 'CategorialAttribute',
-        'format': None}
-        for col in df.select_dtypes('object').columns
-        if col not in bares
-    ]
-
-    dct_bare += [{
-        'name': col,
-        'semantic': 'NumericAttribute',
-        'format': None}
-        for col in df.select_dtypes('number').columns
-        if col not in bares
-    ]
-    dct_bare = [{**{'idx': n}, **dct} for n, dct in enumerate(dct_bare)]
-
-    return df.loc[:, [c['name'] for c in dct_bare]], dct_bare
+    bares = ['Case ID', 'Action']
+    dct_bare = []
+    for i in range(0, len(df.columns)):
+        col = df.columns[i]
+        if pd.api.types.is_object_dtype(df[col]) and col not in bares:
+            dct_bare += [{
+                'name': col,
+                'semantic': 'CategorialAttribute',
+                'format': None,
+                'idx': i}]
+        elif pd.api.types.is_numeric_dtype(df[col]) and col not in bares:
+            dct_bare += [{
+                'name': col,
+                'semantic': 'NumericAttribute',
+                'format': None,
+                'idx': i}]
+        elif col == "Start" or col == "Complete":
+            dct_bare += [{
+                'name': col,
+                'semantic': col,
+                'format': time_format,
+                'idx': i}]
+        else:
+            dct_bare += [{
+                'name': col,
+                'semantic': col,
+                'format': None,
+                'idx': i}]
```

Comment:
Personally, I dislike `if`s with several branches. It always looks much more complex than the task is. Nevertheless, we can make it slightly more readable:

```python
ids = ["Case ID", "Action"]
timestamps = ["Start", "Complete"]

if col in ids:
    # the most general branch, that exists in all logs
    dct_bare += [{
        'name': col,
        'semantic': col,
        'format': None,
        'idx': i}]
elif col in timestamps:
    # the second most general branch, Start exists always, Complete sometimes
    dct_bare += [{
        'name': col,
        'semantic': col,
        'format': time_format,
        'idx': i}]
else:
    # an optional branch, defined by not being (partially) required and thus inferred
    if pd.api.types.is_object_dtype(df[col]):
        dct_bare += [{
            'name': col,
            'semantic': 'CategorialAttribute',
            'format': None,
            'idx': i}]
    elif pd.api.types.is_numeric_dtype(df[col]):
        dct_bare += [{
            'name': col,
            'semantic': 'NumericAttribute',
            'format': None,
            'idx': i}]
```

If we expect to have most of the times less than a couple thousand columns, we can use a comprehension like

```python
bares = ['Case ID', 'Action']
dct_bares = [{
    'name': col,
    'semantic': col,
    'format': None,
    'idx': i}
    for i, col in enumerate(df.columns)
    if col in bares
]
```

for all branches and then combine them into the result.

Reply:
done